_posts/2016-05-12-publish-subscribe-model-in-kafka.md

```diff
@@ -0,0 +1,123 @@
+You must handle Broker leader changes
+
+If you read the code suggested for handling those concerns, you'll be quickly discouraged to use this consumer.
+
+**New Consumer** offers the right level of abstraction and allows us to subscribe to specific partitions. In the documentation they suggest as use case the following:
+
+> The first case is if the process is maintaining some kind of local state associated with that partition (like a local on-disk key-value store) and hence it should only get records for the partition it is maintaining on disk.
+
+Unfortunately our system is using Kafka 0.8, and this consumer is only available from 0.9. We don't have the resources to migrate to that version, so we'll need to stick with **High level consumer**.
+
+That consumer offers a nice API, but it doesn't allow us to subscribe to specific partitions. That means that every instance of the Law enforcement service will consume every message, even those that are not pertinent. We can achieve that defining different consumer groups per instance.
+
+## Leveraging Akka Event Bus
+
+In the [previous post](http://codurance.com/2016/05/10/finite-state-machines-with-akka/) we defined some FSM actor that is waiting for `ItemDeleted` messages.
+
+```scala
+  when(Active) {
+    case Event(ItemDeleted(item), currentItemsToBeDeleted@ItemsToBeDeleted(items)) =>
+      val newItemsToBeDeleted = items.filterNot(_ == item)
+      newItemsToBeDeleted.size match {
+        case 0 => finishWorkWith(CensorResult(Right()))
+        case _ => stay using currentItemsToBeDeleted.copy(items = newItemsToBeDeleted)
+      }
+  }
+```
+
+Our Kafka Consumer could forward every message to those actors and let them to discard/filter not relevant items. However we don't want to overwhelm our actors with redundant and inefficient work, so we'll add a layer of abstraction that will filter the proper messages in a really efficient way.
+
+```scala
+final case class MsgEnvelope(partitionKey: String, payload: ItemDeleted)
+
+class ItemDeletedBus extends EventBus with LookupClassification {
+  override type Event = MsgEnvelope
+  override type Classifier = String
+  override type Subscriber = ActorRef
+
+  override protected def mapSize(): Int = 128
+
+  override protected def publish(event: Event, subscriber: Subscriber): Unit = subscriber ! event.payload
+
+  override protected def classify(event: Event): Classifier = event.partitionKey
+
+  override protected def compareSubscribers(a: Subscriber, b: Subscriber): Int = a.compareTo(b)
+}
+```
+[Akka Event Bus](http://doc.akka.io/docs/akka/2.4.4/scala/event-bus.html) offers us that subscription by partition that we're missing in our Kafka High Level Consumer. From our Kafka Consumer we'll publish every message into the bus:
+
+```scala
+    itemDeletedBus.publish(MsgEnvelope(item.partitionKey, ItemDeleted(item)))
+```
+
+In the previous post we showed how to subscribe to messages for that partition key:
```

Comment:
using that partition key ? 

or 

from that partition topic

is that right to say ?